be/src/storage/lake/horizontal_compaction_task.cpp

```diff
@@ -74,24 +71,40 @@ Status HorizontalCompactionTask::execute(Progress* progress, CancelFunc cancel_f
 #ifndef BE_TEST
         RETURN_IF_ERROR(tls_thread_status.mem_tracker()->check_mem_limit("Compaction"));
 #endif
-        if (auto st = reader.get_next(chunk.get()); st.is_end_of_file()) {
-            break;
-        } else if (!st.ok()) {
-            return st;
+        {
+            SCOPED_RAW_TIMER(&reader_time_ns);
+            if (auto st = reader.get_next(chunk.get()); st.is_end_of_file()) {
+                break;
+            } else if (!st.ok()) {
+                return st;
+            }
         }
         ChunkHelper::padding_char_columns(char_field_indexes, schema, tablet_schema, chunk.get());
         RETURN_IF_ERROR(writer->write(*chunk));
         chunk->reset();
 
-        progress->update(100 * reader.stats().raw_rows_read / total_num_rows);
-        VLOG_EVERY_N(3, 1000) << "Tablet: " << _tablet.id() << ", compaction progress: " << progress->value();
+        _context->progress.update(100 * reader.stats().raw_rows_read / total_num_rows);
+        VLOG_EVERY_N(3, 1000) << "Tablet: " << _tablet.id() << ", compaction progress: " << _context->progress.value();
+        // add reader stats
+        auto stats = reader.stats();
+        _context->stats->reader_time_ns += reader_time_ns;
+        _context->stats->io_ns += stats.io_ns;
+        _context->stats->segment_init_ns += stats.segment_init_ns;
+        _context->stats->column_iterator_init_ns += stats.column_iterator_init_ns;
+        _context->stats->io_count_local_disk += stats.io_count_local_disk;
+        _context->stats->io_count_remote += stats.io_count_remote;
+        _context->stats->compressed_bytes_read += stats.compressed_bytes_read;
```

Comment:
should update stats outside the while loop

Reply:
> should update stats outside the while loop

yes, it's a mistake. done. thanks